healthrex_ml/featurizers/starr_featurizers.py

```diff
@@ -488,3 +488,271 @@ def _build_vocab(self, data):
+
+        # Create working directory if does not already exist and save features
+        os.makedirs(self.outpath, exist_ok=True)
+        save_npz(os.path.join(self.outpath, 'train_features.npz'), train_csr)
+        save_npz(os.path.join(self.outpath, 'test_features.npz'), test_csr)
+        print(f"Feature matrix generated with {train_csr.shape[1]} features")
+
+        # Save labels
+        train_labels.to_csv(os.path.join(self.outpath, 'train_labels.csv'),
+                            index=None)
+        test_labels.to_csv(os.path.join(self.outpath, 'test_labels.csv'),
+                           index=None)
+
+        # Save feature order
+        df_vocab = pd.DataFrame(data={
+            'features': [t for t in train_vocab],
+            'indices': [train_vocab[t] for t in train_vocab]
+        })
+        df_vocab.to_csv(os.path.join(self.outpath, 'feature_order.csv'),
+                        index=None)
+
+        # Save bin thresholds if they exist
+        self.df_lup = pd.DataFrame()
+        for lup in self.lups:
+            if lup is not None:
+                self.df_lup = pd.concat([self.df_lup, lup])
+        if not self.df_lup.empty:
+            self.df_lup.to_csv(os.path.join(self.outpath, 'bin_lup.csv'),
+                               index=None)
+
+        # Save feature_config
+        with open(os.path.join(self.outpath, 'feature_config.json'), 'w') as f:
+            json.dump(self.feature_config, f)
+
+    def construct_feature_timeline(self):
+        """
+        Calls extractors to create long form feature timeline
+        """
+        # Call extractors and collect any look up tables
+        self.lups = []
+        for ext in tqdm(self.extractors):
+            self.lups.append(ext())
+
+    def construct_bag_of_words_rep(self):
+        """
+        Transforms long form feature timeline into a bag of words feature
+        matrix. Stores as a new table in the given bigquery database but also
+        constructs local sparse matrices that can be fed into various sklearn
+        style classifiers. 
+        """
+
+        # Go from timeline to counts
+        query = f"""
+        CREATE OR REPLACE TABLE {self.feature_table_id}_bow AS (
+        SELECT 
+            observation_id, index_time, feature_type, feature, COUNT(*) value 
+        FROM 
+            {self.feature_table_id}
+        WHERE 
+            feature_type IS NOT NULL
+        AND
+            feature IS NOT NULL
+        AND
+            feature NOT LIKE 'NUM__%'
+        GROUP BY 
+            observation_id, index_time, feature_type, feature
+        );
+        
+        INSERT INTO {self.feature_table_id}_bow 
+        SELECT 
+            observation_id, index_time, feature_type, feature, CAST(AVG(value) AS INT64) value 
+        FROM 
+            {self.feature_table_id}
+        WHERE 
+            feature_type IS NOT NULL
+        AND
+            feature IS NOT NULL
+        AND
+            feature LIKE 'NUM__%'
```

Comment:
This is awesome.  How do you handle missing values (ie when no numerical lab result value exists for an example in your dataset)?

Reply:
That's a good point. I didn't handle the missing value cases as I excluded those cases without previous measurements during cohort building. As we're using sparse matric representation, I guess it defaults to zero?

Reply:
Yeah you're right, should just go to zero as written. Fine for now, maybe worth implementing some imputation at some other point down the road.